src/error.rs

```diff
@@ -168,8 +168,78 @@ pub enum ProtocolError {
     MissingRequiredHeader(String),
 
     /// Raised when serializing or de-serializing a message body fails.
-    #[error("serialization error")]
-    BodySerializationError(#[from] serde_json::Error),
+    #[error("ProtocolError: serialization error '{0}'")]
+    BodySerializationError(ContentTypeError),
+}
+
+#[derive(Error, Debug)]
+pub enum ContentTypeError {
+    #[error("'{0}'")]
+    Json(serde_json::Error),
+
+    #[cfg(any(test, feature = "extra_content_types"))]
+    #[error("'{0}'")]
+    Yaml(serde_yaml::Error),
+
+    #[cfg(any(test, feature = "extra_content_types"))]
+    #[error("'{0}'")]
+    Pickle(serde_pickle::error::Error),
+
+    #[cfg(any(test, feature = "extra_content_types"))]
+    #[error("'{0}'")]
+    MsgPackDecode(rmp_serde::decode::Error),
+
+    #[cfg(any(test, feature = "extra_content_types"))]
+    #[error("'{0}'")]
+    MsgPackEncode(rmp_serde::encode::Error),
+
+    #[cfg(any(test, feature = "extra_content_types"))]
+    #[error("'{0}'")]
+    MsgPackValue(rmpv::ext::Error),
+
+    #[error("Unknown content type error")]
+    Unknown,
+}
+
+impl From<serde_json::Error> for ProtocolError {
+    fn from(err: serde_json::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::Json(err))
+    }
+}
+
+#[cfg(any(test, feature = "extra_content_types"))]
+impl From<serde_yaml::Error> for ProtocolError {
+    fn from(err: serde_yaml::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::Yaml(err))
+    }
+}
+
+#[cfg(any(test, feature = "extra_content_types"))]
+impl From<serde_pickle::error::Error> for ProtocolError {
+    fn from(err: serde_pickle::error::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::Pickle(err))
+    }
+}
+
+#[cfg(any(test, feature = "extra_content_types"))]
+impl From<rmp_serde::decode::Error> for ProtocolError {
+    fn from(err: rmp_serde::decode::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::MsgPackDecode(err))
+    }
+}
+
+#[cfg(any(test, feature = "extra_content_types"))]
+impl From<rmp_serde::encode::Error> for ProtocolError {
+    fn from(err: rmp_serde::encode::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::MsgPackEncode(err))
+    }
+}
+
+#[cfg(any(test, feature = "extra_content_types"))]
+impl From<rmpv::ext::Error> for ProtocolError {
+    fn from(err: rmpv::ext::Error) -> Self {
+        Self::BodySerializationError(ContentTypeError::MsgPackValue(err))
+    }
```

Comment:
This part appears to still be necessary, even with the change away from failure. Is there some other way this should be resolved? Without it we cannot use ? all over the place in MessageBuilder::build and Message::body.

Reply:
Hmm, that's strange. It _should_. I'll see if I can figure out what's going on.

Reply:
Oh, right. We're trying to convert `serde_json::Error` directly into `ProtocolError` instead of `ContentTypeError`. I missed that before, sorry!

Reply:
I guess we'll still need those implementations of `From`. But instead of

```rust
impl From<serde_json::Error> for ProtocolError {
    fn from(err: serde_json::Error) -> Self {
        Self::BodySerializationError(ContentTypeError::Json(err))
    }
}
```

for example, we could do

```rust
impl From<serde_json::Error> for ProtocolError {
    fn from(err: serde_json::Error) -> Self {
        Self::from(ContentTypeError::from(err))
    }
}
```

to make use of the existing (derived) `From` implementations.